nixos/modules/services/networking/syncthing.nix

```diff
@@ -51,25 +53,69 @@ let
             --retry 1000 --retry-delay 1 --retry-all-errors \
             "$@"
     }
-
-    # query the old config
-    old_cfg=$(curl ${cfg.guiAddress}/rest/config)
-
-    # generate the new config by merging with the NixOS config options
-    new_cfg=$(printf '%s\n' "$old_cfg" | ${pkgs.jq}/bin/jq -c ${escapeShellArg ''. * ${builtins.toJSON cleanedConfig} * {
-        "devices": ('${escapeShellArg (builtins.toJSON devices)}'${optionalString (cfg.settings.devices == {} || ! cfg.overrideDevices) " + .devices"}),
-        "folders": ('${escapeShellArg (builtins.toJSON folders)}'${optionalString (cfg.settings.folders == {} || ! cfg.overrideFolders) " + .folders"})
-    }''})
-
-    # send the new config
-    curl -X PUT -d "$new_cfg" ${cfg.guiAddress}/rest/config
-
+  '' +
+  /* Syncthing's rest API for the folders and devices is identical. Hence we
+  iterate the cfg.settings.{devices,folder} Nix variables using the
+  conf_type which is either "devices" or "folders": */
+  (lib.concatStringsSep "\n" (map (conf_type: let
+    new_conf_IDs = (concatStringsSep " " {
+      devices = devicesIDs;
+      folders = foldersIDs;
+    }.${conf_type});
+    override = {
+      devices = cfg.overrideDevices;
+      folders = cfg.overrideFolders;
+    }.${conf_type};
+    conf = {
+      inherit devices folders;
+    }.${conf_type};
+    # The API's id naming is slightly different for devices and folders
+    GET_IdAttrName = {
+      devices = "deviceID";
+      folders = "id";
+    }.${conf_type};
+    # All URLs and curl commands are based on: https://docs.syncthing.net/rest/config.html
+    baseAddress = "${cfg.guiAddress}/rest/config/${conf_type}";
+  in ((lib.optionalString (new_conf_IDs != []) ''
+  # We iterate the IDs list, and run a curl -X POST command for each, that
+  # should update that device/folder only.
+  for id in ${new_conf_IDs}; do
+    new_cfg=$(echo ${escapeShellArg (builtins.toJSON conf)} | ${jq} '.[] | select(.${GET_IdAttrName} == "'$id'")')
+    # Quoting https://docs.syncthing.net/rest/config.html: > PUT takes an
```

Comment:
if PUT takes an array, why don't we use this and push an array instead of iterating manually and POSTING each device/folder?

Reply:
> if PUT takes an array, why don't we use this and push an array instead of iterating manually and POSTING each device/folder?

Good question! The answer relates to our support for `override{Folders,Devices}` which complicate things a bit. If you `PUT`, an array, you delete all other folders/devices, and you don't want to do that in case the user doesn't want to override folders/devices. It could have been possible to use `PUT` if `override{Folders,Devices}` is `true`, and iterating a bash loop with `POST` otherwise, but that would have made the logic even less readable, for no real benefit IMO.